clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```
@@ -587,31 +633,57 @@ protected void onJoinPrepare(int generation, String memberId) {
+                // revoke all partitions
+                revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
                 subscriptions.assignFromSubscribed(Collections.emptySet());
 
+                firstException.compareAndSet(null, maybeRevokePartitions(revokedPartitions));
                 break;
 
             case COOPERATIVE:
+                // only revoke those partitions that are not in the subscription any more.
+                Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                revokedPartitions = ownedPartitions.stream()
+                    .filter(tp -> !subscriptions.subscription().contains(tp.topic()))
+                    .collect(Collectors.toSet());
+
+                if (!revokedPartitions.isEmpty()) {
+                    ownedPartitions.removeAll(revokedPartitions);
+                    subscriptions.assignFromSubscribed(ownedPartitions);
+
+                    firstException.compareAndSet(null, maybeRevokePartitions(revokedPartitions));
+                }
+
                 break;
         }
 
         isLeader = false;
         subscriptions.resetGroupSubscription();
+
+        if (firstException.get() != null) {
+            throw new KafkaException("User rebalance callback throws an error", firstException.get());
+        }
+    }
+
+    @Override
+    public void resetGeneration(String errorMessage) {
+        if (subscriptions.partitionsAutoAssigned()) {
+            // lost all partitions if it is based on subscriptions
+            Set<TopicPartition> lostPartitions = new HashSet<>(subscriptions.assignedPartitions());
+            subscriptions.assignFromSubscribed(Collections.emptySet());
+
+            Exception e = maybeLostPartitions(errorMessage, lostPartitions);
+
+            if (e != null) {
+                throw new KafkaException("User rebalance callback throws an error", e);
```

Hmm.. Do we know how this is getting propagated in all cases? Some of the responses are handled using a `RequestCompletionHandler`, but NetworkClient currently eats any exception raised from this callback.

---

Good point, will make a pass through all the call traces.

---

Just to confirm: for AbstractCoordinator we have a specific `CoordinatorResponseHandler` which captures the `RuntimeException` (and KafkaException extends it) and raise to ConsumerCoordinator. So I think that path is covered.